src/java.base/share/classes/sun/security/provider/SecureRandom.java

```diff
@@ -190,25 +204,78 @@ private static void updateState(byte[] state, byte[] output) {
         }
     }
 
+    private void invalidate() {
+        if (state != null) {
+            Arrays.fill(state, (byte)0);
+        }
+        state = null;
+        if (remainder != null) {
+            Arrays.fill(remainder, (byte)0);
+        }
+        remainder = null;
+        remCount = 0;
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        objLock.lock();
+        if (clearStateOnCheckpoint) {
+            invalidate();
+        }
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        objLock.unlock();
+    }
+
+    @Override
+    public Priority getPriority() {
+        return Priority.SECURE_RANDOM;
+    }
+
     /**
      * This static object will be seeded by SeedGenerator, and used
      * to seed future instances of SHA1PRNG SecureRandoms.
      *
      * Bloch, Effective Java Second Edition: Item 71
      */
-    private static class SeederHolder {
-
+    private static class SeederHolder implements jdk.internal.crac.JDKResource {
         private static final SecureRandom seeder;
+        private ReentrantLock objLock = new ReentrantLock();
 
         static {
             /*
              * Call to SeedGenerator.generateSeed() to add additional
              * seed material (likely from the Native implementation).
              */
             seeder = new SecureRandom(SeedGenerator.getSystemEntropy());
-            byte [] b = new byte[DIGEST_SIZE];
+            byte[] b = new byte[DIGEST_SIZE];
             SeedGenerator.generateSeed(b);
             seeder.engineSetSeed(b);
+            jdk.internal.crac.Core.getJDKContext().register(new SeederHolder());
+        }
+
+        @Override
+        public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+            objLock.lock();
```

Comment:
The CheckpointException does not look like a perfect exception here. To avoid controversial requirements, could it be RuntimeException or another unchecked exception, like e.g. IllegalStateException?

Reply:
CheckpointException is CheckpointException now.
We can not use IllegalStateException because we should catch this exception and correctly roll back the checkpoint.
Also, see my comments in CheckpointException https://github.com/openjdk/crac/pull/7#discussion_r793379918